File: java/custom-set/src/main/java/CustomSet.java
import java.util.Collection;

import java.util.ArrayList;
import java.util.List;

class CustomSet<T> {

    private List<T> items;

    CustomSet() {
        this.items = new ArrayList<>();
    }

    CustomSet(Collection<T> data) {
        this();
        data.forEach(this::add);
    }

    boolean isEmpty() {
        return this.items.size() == 0;
    }

    boolean contains(T element) {
        return this.items.contains(element);
    }

    boolean isDisjoint(CustomSet<T> other) {
        return this.items.stream().noneMatch(other::contains);
    }

    boolean add(T element) {
        if (!this.contains(element)) {
            return this.items.add(element);
        }

        return false;
    }

    @Override
    public boolean equals(Object obj) {
        CustomSet<T> other = (CustomSet<T>) obj;

        return other.items.size() == this.items.size()
                && other.items.stream().allMatch(this::contains);
    }

    CustomSet<T> getIntersection(CustomSet<T> other) {
        List<T> data = this.items.stream().filter(other::contains).toList();
        return new CustomSet<T>(data);
    }

    CustomSet<T> getUnion(CustomSet<T> other) {
        CustomSet<T> res = new CustomSet<>(new ArrayList<>(this.items));
        other.items.forEach(res::add);
        
        return res;
    }

    CustomSet<T> getDifference(CustomSet<T> other) {
        List<T> data = this.items.stream().filter(item -> !other.contains(item)).toList();
        return new CustomSet<T>(data);
    }

    boolean isSubset(CustomSet<T> other) {
        if (other.isEmpty()) {
            return true;
        }

        return other.items.stream().allMatch(this::contains);
    }
}

File: python/custom-set/custom_set.py
class CustomSet:
    def __init__(self, elements=[]):
        self.elements = {}
        for element in elements:
            self.elements[element] = 1

    def isempty(self):
        return len(self.elements) == 0

    def __contains__(self, element):
        return self.elements.get(element) == 1

    def issubset(self, other):
        return all(other.__contains__(e) for e in self.elements)

    def isdisjoint(self, other):
        return all(not other.__contains__(e) for e in self.elements)

    def __eq__(self, other):
        return self.issubset(other) and other.issubset(self)

    def add(self, element):
        self.elements[element] = 1

    def intersection(self, other):
        elmnts = [e for e in self.elements if other.__contains__(e)]
        return CustomSet(elmnts)

    def __sub__(self, other):
        elmnts = [e for e in self.elements if not other.__contains__(e)]
        return CustomSet(elmnts)

    def __add__(self, other):
        elmnts = list(self.elements)
        elmnts.extend(list(other.elements))
        return CustomSet(elmnts)
